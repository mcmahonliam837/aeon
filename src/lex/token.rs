use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Identifier(String),
    Keyword(Keyword),
    Operator(Operator),
    Literal(Literal),
    OpenParen,
    CloseParen,
    CloseBrace,
    OpenBrace,
    CloseBracket,
    OpenBracket,
    Comma,
    Dot,
    Newline,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Token::Identifier(s) => write!(f, "{}", s),
            Token::Keyword(k) => write!(f, "{}", k),
            Token::Operator(op) => write!(f, "{}", op),
            Token::Literal(lit) => write!(f, "{}", lit),
            Token::OpenParen => write!(f, "("),
            Token::CloseParen => write!(f, ")"),
            Token::OpenBrace => write!(f, "{{"),
            Token::CloseBrace => write!(f, "}}"),
            Token::OpenBracket => write!(f, "["),
            Token::CloseBracket => write!(f, "]"),
            Token::Comma => write!(f, ","),
            Token::Dot => write!(f, "."),
            Token::Newline => writeln!(f),
        }
    }
}

impl TryFrom<char> for Token {
    type Error = ();
    fn try_from(c: char) -> Result<Self, ()> {
        match c {
            '(' => Ok(Token::OpenParen),
            ')' => Ok(Token::CloseParen),
            '{' => Ok(Token::OpenBrace),
            '}' => Ok(Token::CloseBrace),
            '[' => Ok(Token::OpenBracket),
            ']' => Ok(Token::CloseBracket),
            ',' => Ok(Token::Comma),
            '.' => Ok(Token::Dot),
            _ => Err(()),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Keyword {
    Module,
    Import,
    Fn,
    Struct,
    Enum,
    If,
    Else,
    Return,
    Null,
    Void,
}

impl fmt::Display for Keyword {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Keyword::Module => write!(f, "module"),
            Keyword::Import => write!(f, "import"),
            Keyword::Fn => write!(f, "fn"),
            Keyword::Struct => write!(f, "struct"),
            Keyword::Enum => write!(f, "enum"),
            Keyword::If => write!(f, "if"),
            Keyword::Else => write!(f, "else"),
            Keyword::Return => write!(f, "return"),
            Keyword::Null => write!(f, "null"),
            Keyword::Void => write!(f, "void"),
        }
    }
}

impl TryFrom<&str> for Keyword {
    type Error = ();
    fn try_from(s: &str) -> Result<Self, ()> {
        match s.to_lowercase().as_str() {
            "module" => Ok(Keyword::Module),
            "import" => Ok(Keyword::Import),
            "fn" => Ok(Keyword::Fn),
            "struct" => Ok(Keyword::Struct),
            "enum" => Ok(Keyword::Enum),
            "if" => Ok(Keyword::If),
            "else" => Ok(Keyword::Else),
            "return" => Ok(Keyword::Return),
            "null" => Ok(Keyword::Null),
            "void" => Ok(Keyword::Void),
            _ => Err(()),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Literal {
    String(String),
    Number(String),
    Boolean(bool),
}

impl fmt::Display for Literal {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Literal::String(s) => write!(f, "\"{}\"", s),
            Literal::Number(n) => write!(f, "{}", n),
            Literal::Boolean(b) => write!(f, "{}", b),
        }
    }
}

impl TryFrom<&str> for Literal {
    type Error = ();
    fn try_from(s: &str) -> Result<Self, ()> {
        match s {
            "true" => Ok(Literal::Boolean(true)),
            "false" => Ok(Literal::Boolean(false)),
            s if s.starts_with("0x") => {
                if u64::from_str_radix(&s[2..], 16).is_ok() {
                    Ok(Literal::Number(s.to_string()))
                } else {
                    Err(())
                }
            }
            s if s.starts_with("0b") => {
                if u64::from_str_radix(&s[2..], 2).is_ok() {
                    Ok(Literal::Number(s.to_string()))
                } else {
                    Err(())
                }
            }
            s if s.starts_with("0o") => {
                if u64::from_str_radix(&s[2..], 8).is_ok() {
                    Ok(Literal::Number(s.to_string()))
                } else {
                    Err(())
                }
            }
            s if s.parse::<f64>().is_ok() || s.parse::<u64>().is_ok() => {
                Ok(Literal::Number(s.to_string()))
            }
            _ => Err(()),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Operator {
    Assign,
    Reassign,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Caret,
    Ampersand,
    Pipe,
    Less,
    Greater,
    Equal,
    NotEqual,
    LessEqual,
    GreaterEqual,
    And,
    Or,
    Pipeline,
    Question,
}

impl fmt::Display for Operator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Operator::Assign => write!(f, ":"),
            Operator::Reassign => write!(f, "="),
            Operator::Plus => write!(f, "+"),
            Operator::Minus => write!(f, "-"),
            Operator::Star => write!(f, "*"),
            Operator::Slash => write!(f, "/"),
            Operator::Percent => write!(f, "%"),
            Operator::Caret => write!(f, "^"),
            Operator::Ampersand => write!(f, "&"),
            Operator::Pipe => write!(f, "|"),
            Operator::Less => write!(f, "<"),
            Operator::Greater => write!(f, ">"),
            Operator::Equal => write!(f, "=="),
            Operator::NotEqual => write!(f, "!="),
            Operator::LessEqual => write!(f, "<="),
            Operator::GreaterEqual => write!(f, ">="),
            Operator::And => write!(f, "&&"),
            Operator::Or => write!(f, "||"),
            Operator::Pipeline => write!(f, "|>"),
            Operator::Question => write!(f, "?"),
        }
    }
}

impl TryFrom<&str> for Operator {
    type Error = ();
    fn try_from(s: &str) -> Result<Self, ()> {
        match s {
            ":" => Ok(Operator::Assign),
            "=" => Ok(Operator::Reassign),
            "+" => Ok(Operator::Plus),
            "-" => Ok(Operator::Minus),
            "*" => Ok(Operator::Star),
            "/" => Ok(Operator::Slash),
            "%" => Ok(Operator::Percent),
            "^" => Ok(Operator::Caret),
            "&" => Ok(Operator::Ampersand),
            "|" => Ok(Operator::Pipe),

            "<" => Ok(Operator::Less),
            ">" => Ok(Operator::Greater),

            "&&" => Ok(Operator::And),
            "||" => Ok(Operator::Or),

            "!=" => Ok(Operator::NotEqual),
            "==" => Ok(Operator::Equal),
            "<=" => Ok(Operator::LessEqual),
            ">=" => Ok(Operator::GreaterEqual),
            "|>" => Ok(Operator::Pipeline),
            "?" => Ok(Operator::Question),
            _ => Err(()),
        }
    }
}
