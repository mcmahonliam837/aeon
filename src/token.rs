#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Identifier(String),
    Keyword(Keyword),
    Operator(Operator),
    LiteralString(String),
    LiteralInteger(usize),
    LiteralFloat(f64),
    OpenParenthesis,
    CloseParenthesis,
    CloseBrace,
    OpenBrace,
    CloseBracket,
    OpenBracket,
    Comma,
    Dot,
}

impl TryFrom<char> for Token {
    type Error = ();
    fn try_from(c: char) -> Result<Self, ()> {
        match c {
            '(' => Ok(Token::OpenParenthesis),
            ')' => Ok(Token::CloseParenthesis),
            '{' => Ok(Token::OpenBrace),
            '}' => Ok(Token::CloseBrace),
            '[' => Ok(Token::OpenBracket),
            ']' => Ok(Token::CloseBracket),
            ',' => Ok(Token::Comma),
            '.' => Ok(Token::Dot),
            _ => Err(()),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Keyword {
    Module,
    Import,
    Fn,
    If,
    Else,
    Return,
}

impl TryFrom<&str> for Keyword {
    type Error = ();
    fn try_from(s: &str) -> Result<Self, ()> {
        match s.to_lowercase().as_str() {
            "module" => Ok(Keyword::Module),
            "import" => Ok(Keyword::Import),
            "fn" => Ok(Keyword::Fn),
            "if" => Ok(Keyword::If),
            "else" => Ok(Keyword::Else),
            "return" => Ok(Keyword::Return),
            _ => Err(()),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Literal {
    String(String),
    Integer(usize),
    Float(f64),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Operator {
    Assign,
    Reassign,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Caret,
    Ampersand,
    Pipe,
    Less,
    Greater,
    Equal,
    NotEqual,
    LessEqual,
    GreaterEqual,
    And,
    Or,
}

impl TryFrom<&str> for Operator {
    type Error = ();
    fn try_from(s: &str) -> Result<Self, ()> {
        match s {
            ":=" => Ok(Operator::Assign),
            "=" => Ok(Operator::Reassign),
            "+" => Ok(Operator::Plus),
            "-" => Ok(Operator::Minus),
            "*" => Ok(Operator::Star),
            "/" => Ok(Operator::Slash),
            "%" => Ok(Operator::Percent),
            "^" => Ok(Operator::Caret),
            "&" => Ok(Operator::Ampersand),
            "|" => Ok(Operator::Pipe),

            "<" => Ok(Operator::Less),
            ">" => Ok(Operator::Greater),

            "&&" => Ok(Operator::And),
            "||" => Ok(Operator::Or),

            "!=" => Ok(Operator::NotEqual),
            "==" => Ok(Operator::Equal),
            "<=" => Ok(Operator::LessEqual),
            ">=" => Ok(Operator::GreaterEqual),
            _ => Err(()),
        }
    }
}
