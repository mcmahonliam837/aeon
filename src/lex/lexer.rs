use crate::lex::token::{Keyword, Operator, Token};
use std::{collections::VecDeque, error::Error, io::BufRead};
use utf8_chars::BufReadCharsExt;

#[derive(Debug)]
pub enum LexerError {
    IoError(std::io::Error),
}

impl std::fmt::Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LexerError::IoError(err) => write!(f, "IO error: {}", err),
        }
    }
}

impl Error for LexerError {}

impl From<std::io::Error> for LexerError {
    fn from(err: std::io::Error) -> Self {
        LexerError::IoError(err)
    }
}

enum LexerState {
    InString,
    InComment,
}

enum PostProcessingCommand {
    Clear,
    ClearAndSkipPeak,
}

pub struct Lexer<R: BufRead> {
    reader: R,
}

struct LexerContext {
    current_string: String,
    tokens: Vec<Token>,
    state: VecDeque<LexerState>,
}

impl<R: BufRead> Lexer<R> {
    pub fn lex(reader: R) -> Result<Vec<Token>, LexerError> {
        let context = LexerContext {
            current_string: String::new(),
            tokens: Vec::new(),
            state: VecDeque::new(),
        };
        let mut lexer = Self { reader: reader };
        lexer.run(context)
    }

    fn run(&mut self, mut context: LexerContext) -> Result<Vec<Token>, LexerError> {
        let mut chars = self.reader.chars().peekable();

        while let Some(Ok(c)) = chars.next() {
            let peak = chars.peek().and_then(|r| r.as_ref().ok()).copied();
            let Some(cmd) = Self::process_byte(&mut context, c, peak) else {
                continue;
            };

            context.current_string.clear();
            if matches!(cmd, PostProcessingCommand::ClearAndSkipPeak) {
                chars.next();
            }
        }
        Ok(context.tokens)
    }

    fn process_byte(
        context: &mut LexerContext,
        c: char,
        peak: Option<char>,
    ) -> Option<PostProcessingCommand> {
        match context.state.back() {
            Some(LexerState::InString) => Self::process_string(context, c),
            Some(LexerState::InComment) => {
                if c == '\n' {
                    context.state.pop_back();
                }
                None
            }
            None => Self::process_normal(context, c, peak),
        }
    }

    fn process_string(context: &mut LexerContext, c: char) -> Option<PostProcessingCommand> {
        if c == '"' && !context.current_string.ends_with('\\') {
            context.state.pop_back();
            context
                .tokens
                .push(Token::LiteralString(context.current_string.clone()));
            Some(PostProcessingCommand::Clear)
        } else {
            context.current_string.push(c);
            None
        }
    }

    fn process_normal(
        context: &mut LexerContext,
        c: char,
        peak: Option<char>,
    ) -> Option<PostProcessingCommand> {
        match c {
            '\n' => {
                context.tokens.push(Token::Newline);
                Some(PostProcessingCommand::Clear)
            }
            '"' => {
                context.state.push_back(LexerState::InString);
                None
            }
            '/' if peak == Some('/') => {
                context.state.push_back(LexerState::InComment);
                Self::commit_identifier(context);
                None
            }
            c if c.is_whitespace() => {
                if !context.current_string.is_empty() {
                    Self::commit_identifier(context);
                    Some(PostProcessingCommand::Clear)
                } else {
                    None
                }
            }
            c => Self::process_operator_or_char(context, c, peak),
        }
    }

    fn process_operator_or_char(
        context: &mut LexerContext,
        c: char,
        peak: Option<char>,
    ) -> Option<PostProcessingCommand> {
        if let Ok(token) = Token::try_from(c) {
            Self::commit_and_push(context, token);
            return Some(PostProcessingCommand::Clear);
        }

        let two_char = format!("{}{}", c, peak.unwrap_or(' '));
        if let Ok(op) = Operator::try_from(two_char.as_str()) {
            Self::commit_and_push(context, Token::Operator(op));
            return Some(PostProcessingCommand::ClearAndSkipPeak);
        }

        if let Ok(op) = Operator::try_from(c.to_string().as_str()) {
            Self::commit_and_push(context, Token::Operator(op));
            return Some(PostProcessingCommand::Clear);
        }

        context.current_string.push(c);
        None
    }

    fn commit_and_push(context: &mut LexerContext, token: Token) {
        Self::commit_identifier(context);
        context.tokens.push(token);
    }

    fn commit_identifier(context: &mut LexerContext) {
        if context.current_string.is_empty() {
            return;
        }

        let token = if let Ok(keyword) = Keyword::try_from(context.current_string.as_str()) {
            Token::Keyword(keyword)
        } else {
            Token::Identifier(context.current_string.clone())
        };
        context.tokens.push(token);
        context.current_string.clear();
    }
}
