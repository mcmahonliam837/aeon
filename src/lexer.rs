use crate::token::{Keyword, Operator, Token};
use std::{collections::VecDeque, error::Error, io::BufReader};
use utf8_chars::BufReadCharsExt;

#[derive(Debug)]
pub enum LexerError {
    IoError(std::io::Error),
}

impl std::fmt::Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LexerError::IoError(err) => write!(f, "IO error: {}", err),
        }
    }
}

impl Error for LexerError {}

impl From<std::io::Error> for LexerError {
    fn from(err: std::io::Error) -> Self {
        LexerError::IoError(err)
    }
}

enum LexerState {
    InString,
    InComment,
}

enum PostProcessingCommand {
    Clear,
    ClearAndSkipPeak,
}

pub struct Lexer {
    file_path: String,
    current_string: String,
    tokens: Vec<Token>,
    state: VecDeque<LexerState>,
}

impl Lexer {
    pub fn lex(file_path: &str) -> Result<Vec<Token>, LexerError> {
        let mut lexer = Lexer::new(file_path);
        lexer.run()?;
        Ok(lexer.tokens)
    }

    fn new(file_name: &str) -> Self {
        Self {
            file_path: String::from(file_name),
            current_string: String::new(),
            tokens: Vec::new(),
            state: VecDeque::new(),
        }
    }

    fn run(&mut self) -> Result<(), LexerError> {
        let input = std::fs::File::open(self.file_path.clone())?;
        let mut reader = BufReader::new(input);

        let mut position = 0;
        let mut tmp = reader.chars().peekable();
        while let Some(c) = tmp.next() {
            match c {
                Ok(c) => {
                    let peak = tmp.peek().and_then(|r| r.as_ref().ok()).copied();
                    let Some(command) = self.process_byte(c, peak) else {
                        continue;
                    };
                    match command {
                        PostProcessingCommand::Clear => {
                            self.current_string.clear();
                        }
                        PostProcessingCommand::ClearAndSkipPeak => {
                            self.current_string.clear();
                            tmp.next();
                        }
                    }
                }
                Err(ref err) if err.kind() == std::io::ErrorKind::UnexpectedEof => {
                    break;
                }
                Err(err) => {
                    return Err(LexerError::IoError(err));
                }
            }
            position += 1;
        }

        Ok(())
    }

    fn process_byte(&mut self, c: char, peak: Option<char>) -> Option<PostProcessingCommand> {
        match self.state.back() {
            None => match c {
                '\n' => {
                    self.tokens.push(Token::Newline);
                    Some(PostProcessingCommand::Clear)
                }
                '"' => {
                    self.state.push_back(LexerState::InString);
                    None
                }
                '/' if peak == Some('/') => {
                    self.state.push_back(LexerState::InComment);
                    if !self.current_string.is_empty() {
                        let token = Self::process_unknown_string(self.current_string.as_str());
                        self.tokens.push(token);
                    }
                    self.current_string.clear();
                    None
                }
                c if let Ok(token) = Token::try_from(c) => {
                    Self::commit_and_push(&mut self.tokens, token, self.current_string.as_str());
                    Some(PostProcessingCommand::Clear)
                }
                c if c.is_whitespace() && !self.current_string.is_empty() => {
                    if let Ok(keyword) = Keyword::try_from(self.current_string.as_str()) {
                        self.tokens.push(Token::Keyword(keyword));
                    } else {
                        self.tokens
                            .push(Token::Identifier(self.current_string.clone()));
                    }
                    Some(PostProcessingCommand::Clear)
                }
                c if c.is_whitespace() => None,
                c => {
                    let mut operator_chars = vec![c];
                    if let Some(peak) = peak {
                        operator_chars.push(peak);
                    };

                    let op_str = String::from_iter(operator_chars);

                    if let Ok(op) = Operator::try_from(op_str.as_str()) {
                        Self::commit_and_push(
                            &mut self.tokens,
                            Token::Operator(op),
                            self.current_string.as_str(),
                        );
                        return Some(PostProcessingCommand::ClearAndSkipPeak);
                    }

                    if let Ok(op) = Operator::try_from(c.to_string().as_str()) {
                        Self::commit_and_push(
                            &mut self.tokens,
                            Token::Operator(op),
                            self.current_string.as_str(),
                        );

                        return Some(PostProcessingCommand::ClearAndSkipPeak);
                    }

                    self.current_string.push(c);
                    None
                }
            },
            Some(LexerState::InString) => match c {
                '"' if !self.current_string.ends_with("\\") => {
                    self.state.pop_back();
                    self.tokens
                        .push(Token::LiteralString(self.current_string.clone()));
                    Some(PostProcessingCommand::Clear)
                }
                _ => {
                    self.current_string.push(c);
                    None
                }
            },
            Some(LexerState::InComment) => match c {
                '\n' => {
                    self.state.pop_back();
                    None
                }
                _ => None,
            },
        }
    }

    fn commit_and_push(tokens: &mut Vec<Token>, token: Token, str: &str) {
        if !str.is_empty() {
            tokens.push(Self::process_unknown_string(str));
        }
        tokens.push(token);
    }

    fn process_unknown_string(str: &str) -> Token {
        if let Ok(keyword) = Keyword::try_from(str) {
            Token::Keyword(keyword)
        } else {
            Token::Identifier(str.to_string())
        }
    }
}
