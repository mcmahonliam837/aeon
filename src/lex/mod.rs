pub mod lexer;
pub mod token;

#[cfg(test)]
mod tests {
    use std::io::BufReader;

    use insta::assert_debug_snapshot;

    use stringreader::StringReader;

    use crate::lex::token::Token;

    use super::lexer::*;

    fn lex(path: &str) -> Result<Vec<Token>, LexerError> {
        let source = std::fs::read_to_string(path).expect("failed to load test source");
        let reader = StringReader::new(source.as_str());
        Lexer::lex(BufReader::new(reader))
    }

    #[test]
    fn test_hello_world() {
        assert_debug_snapshot!(lex("aeon_examples/hello_world.aeon"));
    }

    #[test]
    fn test_modules() {
        assert_debug_snapshot!(lex("aeon_examples/modules.aeon"));
    }
}
