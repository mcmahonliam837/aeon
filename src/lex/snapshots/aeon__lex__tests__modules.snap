---
source: src/lex/mod.rs
expression: "lex(\"aeon_examples/modules.aeon\")"
---
Ok(
    [
        Keyword(
            Module,
        ),
        Identifier(
            "Main",
        ),
        Newline,
        Keyword(
            Import,
        ),
        Literal(
            String(
                "std/io",
            ),
        ),
        Newline,
        Keyword(
            Fn,
        ),
        Identifier(
            "main",
        ),
        OpenParen,
        CloseParen,
        OpenBrace,
        Identifier(
            "list",
        ),
        Operator(
            Assign,
        ),
        Operator(
            Reassign,
        ),
        Identifier(
            "LinkedList",
        ),
        Dot,
        Identifier(
            "new",
        ),
        OpenParen,
        CloseParen,
        Dot,
        Identifier(
            "push",
        ),
        OpenParen,
        Literal(
            Number(
                "0",
            ),
        ),
        CloseParen,
        Newline,
        Identifier(
            "io",
        ),
        Dot,
        Identifier(
            "printf",
        ),
        OpenParen,
        Literal(
            String(
                "%A\\n",
            ),
        ),
        Comma,
        Identifier(
            "list",
        ),
        CloseParen,
        Newline,
        OpenBrace,
        Identifier(
            "list",
        ),
        Operator(
            Assign,
        ),
        Keyword(
            Mut,
        ),
        Operator(
            Reassign,
        ),
        Identifier(
            "list",
        ),
        Newline,
        Dot,
        Identifier(
            "push",
        ),
        OpenParen,
        Literal(
            Number(
                "1",
            ),
        ),
        CloseParen,
        Newline,
        Dot,
        Identifier(
            "push",
        ),
        OpenParen,
        Literal(
            Number(
                "2",
            ),
        ),
        CloseParen,
        Newline,
        Dot,
        Identifier(
            "push",
        ),
        OpenParen,
        Literal(
            Number(
                "3",
            ),
        ),
        CloseParen,
        Newline,
        Identifier(
            "list",
        ),
        Dot,
        Identifier(
            "push_mut",
        ),
        OpenParen,
        Literal(
            Number(
                "4",
            ),
        ),
        CloseParen,
        Newline,
        Identifier(
            "io",
        ),
        Dot,
        Identifier(
            "printf",
        ),
        OpenParen,
        Literal(
            String(
                "%A\\n",
            ),
        ),
        Comma,
        Identifier(
            "list",
        ),
        CloseParen,
        Newline,
        CloseBrace,
        Newline,
        Identifier(
            "io",
        ),
        Dot,
        Identifier(
            "printf",
        ),
        OpenParen,
        Literal(
            String(
                "%A\\n",
            ),
        ),
        Comma,
        Identifier(
            "list",
        ),
        CloseParen,
        Newline,
        CloseBrace,
        Newline,
        Keyword(
            Module,
        ),
        Identifier(
            "LinkedList",
        ),
        OpenBrace,
        Keyword(
            Module,
        ),
        Identifier(
            "Node",
        ),
        OpenBrace,
        Keyword(
            Struct,
        ),
        OpenBrace,
        Identifier(
            "data",
        ),
        Operator(
            Assign,
        ),
        Identifier(
            "i32",
        ),
        Newline,
        Identifier(
            "next",
        ),
        Operator(
            Assign,
        ),
        Operator(
            Question,
        ),
        Operator(
            Star,
        ),
        Identifier(
            "Node",
        ),
        Newline,
        CloseBrace,
        Newline,
        CloseBrace,
        Newline,
        Keyword(
            Struct,
        ),
        OpenBrace,
        Identifier(
            "root",
        ),
        Operator(
            Assign,
        ),
        Operator(
            Question,
        ),
        Operator(
            Star,
        ),
        Identifier(
            "Node",
        ),
        Newline,
        CloseBrace,
        Newline,
        Keyword(
            Fn,
        ),
        Identifier(
            "new",
        ),
        OpenParen,
        CloseParen,
        Identifier(
            "Self",
        ),
        OpenBrace,
        Keyword(
            Return,
        ),
        Identifier(
            "LinkedList",
        ),
        OpenBrace,
        Identifier(
            "root",
        ),
        Operator(
            Assign,
        ),
        Keyword(
            Null,
        ),
        CloseBrace,
        Newline,
        CloseBrace,
        Newline,
        Keyword(
            Fn,
        ),
        Identifier(
            "push",
        ),
        OpenParen,
        Identifier(
            "self",
        ),
        Comma,
        Identifier(
            "item",
        ),
        Operator(
            Assign,
        ),
        Identifier(
            "i32",
        ),
        CloseParen,
        Identifier(
            "Self",
        ),
        OpenBrace,
        Keyword(
            Return,
        ),
        Identifier(
            "Self",
        ),
        OpenBrace,
        Identifier(
            "root",
        ),
        Operator(
            Assign,
        ),
        Operator(
            Ampersand,
        ),
        Identifier(
            "Node",
        ),
        OpenBrace,
        Identifier(
            "data",
        ),
        Operator(
            Assign,
        ),
        Identifier(
            "item",
        ),
        Newline,
        Identifier(
            "next",
        ),
        Operator(
            Assign,
        ),
        Identifier(
            "self",
        ),
        Dot,
        Identifier(
            "root",
        ),
        Newline,
        CloseBrace,
        Newline,
        CloseBrace,
        Newline,
        CloseBrace,
        Newline,
        Keyword(
            Fn,
        ),
        Identifier(
            "push_mut",
        ),
        OpenParen,
        Identifier(
            "self",
        ),
        Comma,
        Identifier(
            "item",
        ),
        Operator(
            Assign,
        ),
        Identifier(
            "i32",
        ),
        CloseParen,
        OpenBrace,
        Identifier(
            "new_node",
        ),
        Operator(
            Reassign,
        ),
        Operator(
            Ampersand,
        ),
        Identifier(
            "Node",
        ),
        OpenBrace,
        Identifier(
            "data",
        ),
        Operator(
            Assign,
        ),
        Identifier(
            "item",
        ),
        Newline,
        Identifier(
            "next",
        ),
        Operator(
            Assign,
        ),
        Identifier(
            "self",
        ),
        Dot,
        Identifier(
            "root",
        ),
        Newline,
        CloseBrace,
        Newline,
        Identifier(
            "self",
        ),
        Dot,
        Identifier(
            "root",
        ),
        Operator(
            Reassign,
        ),
        Identifier(
            "new_node",
        ),
        Newline,
        CloseBrace,
        Newline,
        CloseBrace,
        Newline,
    ],
)
